src/CoreWCF.Queue/src/CoreWCF/Queue/Common/QueuePollingService.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using CoreWCF.Channels;
+using CoreWCF.Configuration;
+using CoreWCF.Queue.Common.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace CoreWCF.Queue.Common
+{
+    public class QueuePollingService : IHostedService
+    {
+        private readonly QueueHandShakeMiddleWare _queueHandShakeMiddleWare;
+        private readonly IServiceProvider _services;
+        private readonly List<QueueTransportContext> _queueTransportContexts;
+        private readonly IOptions<QueueOptions> _options;
+        private readonly IServiceBuilder _serviceBuilder;
+        private bool _isServiceBuilderOpened;
+
+        public QueuePollingService(IServiceProvider services, QueueHandShakeMiddleWare queueHandShakeMiddle,
+            IOptions<QueueOptions> queueOptions)
+        {
+            _services = services;
+            _queueHandShakeMiddleWare = queueHandShakeMiddle;
+            _options = queueOptions;
+            _serviceBuilder = _services.GetRequiredService<IServiceBuilder>();
+            _queueTransportContexts = new List<QueueTransportContext>();
+            _serviceBuilder.Opened += (_, _) => _isServiceBuilderOpened = true;
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            Task.Run(() => WaitAndStart(cancellationToken), cancellationToken);
+            return Task.CompletedTask;
+        }
+
+        private async Task WaitAndStart(CancellationToken cancellationToken)
+        {
+            while (!_isServiceBuilderOpened)
+            {
+                await Task.Delay(100, cancellationToken);
+            }
```

Comment:
Yes, create a TaskCompletionSource in the constructor and have the Opened event handler complete it. In this method first check the state of the service builder and if not Opened, await the TaskCompletionSource.Task. You don't want wait unconditionally as there is a potential race condition when the service builder is being opened where it might get past the point of calling the event handler before you set it. So only await the Task if it's not opened. If you want to honor the cancellationToken, register a cancellation callback with the token which cancels the TaskCompletionSource (don't forget to dispose the registration).